cmd: fix race in fsDeleteFile, make isDirEmpty deterministic, drop stat in deleteFile, DRY

brendanashworth opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
```
commit fdbb13799b411b539fca7d9e5b9c05211da2545d
Author: Brendan Ashworth <brendan.ashworth@me.com>
Date:   Mon Jul 31 15:54:22 2017 -0700

    posix: do not upstream errors in deleteFile
    
    This commit changes posix's deleteFile() to not upstream errors from
    removing parent directories. This fixes a race condition.
    
    The race condition occurs when multiple deleteFile()s are called on the
    same parent directory, but different child files. Because deleteFile()
    recursively removes parent directories if they are empty, but
    deleteFile() errors if the selected deletePath does not exist, there was
    an opportunity for a race condition. The two processes would remove the
    child directories successfully, then depend on the parent directory
    still existing. In some cases this is an invalid assumption, because
    other processes can remove the parent directory beforehand. This commit
    changes deleteFile() to not upstream an error if one occurs, because the
    only required error should be from the immediate deletePath, not from a
    parent path.
    
    In the specific bug report, multiple CompleteMultipartUpload requests
    would launch multiple deleteFile() requests. Because they chain up on
    parent directories, ultimately at the end, there would be multiple
    remove files for the ultimate parent directory,
    .minio.sys/multipart/{bucket}. Because only one will succeed and one
    will fail, an error would be upstreamed saying that the file does not
    exist, and the CompleteMultipartUpload code interpreted this as
    NoSuchKey, or that the object/part id doesn't exist. This was faulty
    behavior and is now fixed.
    
    Fixes: https://github.com/minio/minio/issues/4727

commit 52e058583012560c1d7b26a191ee904e422fbb5f
Author: Brendan Ashworth <brendan.ashworth@me.com>
Date:   Fri Jul 28 15:34:01 2017 -0700

    fs: drop Stat() call from fsDeleteFile,deleteFile
    
    This commit makes fsDeleteFile() simply call deleteFile() after calling
    the relevant path length checking functions. This DRYs the code base.
    
    This commit removes the Stat() call from deleteFile(). This improves
    performance and removes any possibility of a race condition.
    
    This additionally adds tests and a benchmark for said function. Here are
    the results from before and after this commit:
    
    benchmark                   old ns/op     new ns/op     delta
    BenchmarkFSDeleteFile-4     2694884       1900661       -29.47%

commit 1778101f5e830f0afea18de67d5c8ce10e4af972
Author: Brendan Ashworth <brendan.ashworth@me.com>
Date:   Thu Jul 20 17:13:15 2017 -0700

    posix: test isDirEmpty, change error conditional
    
    This commit adds a new test for isDirEmpty (for code coverage) and
    changes around the error conditional. Previously, there was a `return
    nil` statement that would only be triggered under a race condition and
    would trip up our test coverage for no reason. With this new error
    conditional, there's no awkward 'else'-esque condition, which means test
    coverage will not change between runs for no reason in this specific
    test. It's also a cleaner read.
```

## Motivation and Context
Fixes the race condition in https://github.com/minio/minio/issues/4727.

Fixes `posix.go` and `fs-v1-helpers.go` in https://github.com/minio/minio/issues/4658#issuecomment-318495587.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
well

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

brendanashworth committed:
posix: test isDirEmpty, change error conditional

This commit adds a new test for isDirEmpty (for code coverage) and
changes around the error conditional. Previously, there was a `return
nil` statement that would only be triggered under a race condition and
would trip up our test coverage for no reason. With this new error
conditional, there's no awkward 'else'-esque condition, which means test
coverage will not change between runs for no reason in this specific
test. It's also a cleaner read.

brendanashworth committed:
fs: drop Stat() call from fsDeleteFile,deleteFile

This commit makes fsDeleteFile() simply call deleteFile() after calling
the relevant path length checking functions. This DRYs the code base.

This commit removes the Stat() call from deleteFile(). This improves
performance and removes any possibility of a race condition.

This additionally adds tests and a benchmark for said function. Here are
the results from before and after this commit:

benchmark                   old ns/op     new ns/op     delta
BenchmarkFSDeleteFile-4     2694884       1900661       -29.47%

harshavardhana commented on cmd/fs-v1-helpers.go:
```diff
@@ -365,6 +365,7 @@ func fsRenameFile(sourcePath, destPath string) error {
 
 // Delete a file and its parent if it is empty at the destination path.
 // this function additionally protects the basePath from being deleted.
+// This will return nil if called on a non-empty directory.
```
We need to talk about this.. can you explain the reasoning? @brendanashworth

brendanashworth replied:
Yes, this isn't actually a change in behavior. I added this comment to clarify what the function does (hence the confusion 😄 ), as [before](https://github.com/minio/minio/blob/ec5293ce29ef03726c7174d9e3304bc156259c4e/cmd/fs-v1-helpers.go#L392-L395) it would `return nil` from the `Stat()` call. This actually led to an interesting race condition (never became an issue afaik), because if `Stat()` returned a non-empty directory, but was empty by the time the `Remove()` call came around, [it would error](https://github.com/minio/minio/blob/ec5293ce29ef03726c7174d9e3304bc156259c4e/cmd/fs-v1-helpers.go#L404). So technically this PR fixes that race condition, even though I doubt it would ever come to fruition in a production environment.

Also, all tests that are written for this commit pass both before and after the changes. This was done to prevent any change in behavior.

brendanashworth committed:
posix: do not upstream errors in deleteFile

This commit changes posix's deleteFile() to not upstream errors from
removing parent directories. This fixes a race condition.

The race condition occurs when multiple deleteFile()s are called on the
same parent directory, but different child files. Because deleteFile()
recursively removes parent directories if they are empty, but
deleteFile() errors if the selected deletePath does not exist, there was
an opportunity for a race condition. The two processes would remove the
child directories successfully, then depend on the parent directory
still existing. In some cases this is an invalid assumption, because
other processes can remove the parent directory beforehand. This commit
changes deleteFile() to not upstream an error if one occurs, because the
only required error should be from the immediate deletePath, not from a
parent path.

In the specific bug report, multiple CompleteMultipartUpload requests
would launch multiple deleteFile() requests. Because they chain up on
parent directories, ultimately at the end, there would be multiple
remove files for the ultimate parent directory,
.minio.sys/multipart/{bucket}. Because only one will succeed and one
will fail, an error would be upstreamed saying that the file does not
exist, and the CompleteMultipartUpload code interpreted this as
NoSuchKey, or that the object/part id doesn't exist. This was faulty
behavior and is now fixed.

Fixes: https://github.com/minio/minio/issues/4727
